narwhal/executor/src/lib.rs

```diff
@@ -112,20 +114,40 @@ pub async fn get_restored_consensus_output<State: ExecutionState>(
         .read_last_consensus_index()
         .map_err(SubscriberError::StoreError)?;
 
-    let next_cert_index = execution_state
+    // Execution state always keeps the index of the latest certificate that has been executed.
+    // However, in consensus_store the committed certificates are stored alongside the "next" index
+    // that should be assigned to a certificate. Thus, to successfully recover the un-executed
+    // certificates from the consensus_store we are incrementing the `next_certificate_index` by 1
+    // to align the semantics.
+    // TODO: https://github.com/MystenLabs/sui/issues/5819
```

Comment:
Added this ticket as place holder. I would like us to refactor the `next_certificate_index` semantics as it's super confusing the fact that we always store alongside not the index of the certificate that we just committed, but the "next" one.